pkg/controller/deployment/sync.go

```diff
@@ -101,37 +101,39 @@ func (dc *DeploymentController) rsAndPodsWithHashKeySynced(deployment *extension
 		return nil, nil, fmt.Errorf("error listing ReplicaSets: %v", err)
 	}
 	syncedRSList := []*extensions.ReplicaSet{}
+	syncedPodList := &api.PodList{Items: []api.Pod{}}
 	for _, rs := range rsList {
 		// Add pod-template-hash information if it's not in the RS.
 		// Otherwise, new RS produced by Deployment will overlap with pre-existing ones
 		// that aren't constrained by the pod-template-hash.
-		syncedRS, err := dc.addHashKeyToRSAndPods(rs)
+		syncedRS, syncedPods, err := dc.addHashKeyToRSAndPods(rs)
 		if err != nil {
 			return nil, nil, err
 		}
 		syncedRSList = append(syncedRSList, syncedRS)
+		if syncedPods != nil {
+			syncedPodList.Items = append(syncedPodList.Items, syncedPods.Items...)
+		}
 	}
-	syncedPodList, err := dc.listPods(deployment)
-	if err != nil {
-		return nil, nil, err
-	}
+	// if we failed to get list of pods when adding hash but we got syncedRSList, attempt to
+	// list the pods for the deployment.
 	return syncedRSList, syncedPodList, nil
 }
 
 // addHashKeyToRSAndPods adds pod-template-hash information to the given rs, if it's not already there, with the following steps:
 // 1. Add hash label to the rs's pod template, and make sure the controller sees this update so that no orphaned pods will be created
 // 2. Add hash label to all pods this rs owns, wait until replicaset controller reports rs.Status.FullyLabeledReplicas equal to the desired number of replicas
 // 3. Add hash label to the rs's label and selector
-func (dc *DeploymentController) addHashKeyToRSAndPods(rs *extensions.ReplicaSet) (updatedRS *extensions.ReplicaSet, err error) {
+func (dc *DeploymentController) addHashKeyToRSAndPods(rs *extensions.ReplicaSet) (*extensions.ReplicaSet, *api.PodList, error) {
 	objCopy, err := api.Scheme.Copy(rs)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	updatedRS = objCopy.(*extensions.ReplicaSet)
+	updatedRS := objCopy.(*extensions.ReplicaSet)
 
 	// If the rs already has the new hash label in its selector, it's done syncing
 	if labelsutil.SelectorHasLabel(rs.Spec.Selector, extensions.DefaultDeploymentUniqueLabelKey) {
-		return
+		return nil, nil, nil
```

Comment:
Looking closer at the code, it turns out we need to list pods twice atm. If you return no pods here for one of the old replica sets that still has pods, then `rsAndPodsWithHashKeySynced` will not return all the pods it can list, and its caller (`getAllReplicaSetsAndSyncRevision`) by using `FindOldReplicaSets` will not return all the active replica sets. In reality, we don't need to return pods at all from `rsAndPodsWithHashKeySynced` and `FindOldReplicaSets` is redundant because it tries to identify the active replica sets while we should already have that info in the replica set status.